pwiz_tools/Skyline/Util/ParsedMolecule.cs

```diff
@@ -204,6 +206,90 @@ public static bool TryParseFormula(string formula, out ParsedMolecule resultMole
+        {
+            modMassMono = null;
+            modMassAverage = null;
+            formula = formulaAndMasses?.Trim();
+            if (string.IsNullOrEmpty(formula))
+            {
+                return;
+            }
+            // A few different possibilities here, e.g. "C12H5", "C12H5[-1.2/1.21]", "[+1.3/1.31]", "1.3/1.31"
+            // Also possibly  "C12H5[-1.2/1.21]-C2H[-1.1]"
+            var position = 0;
+            var success = true;
+            while (success && MoleculeMassOffset.StringContainsMassOffsetCue(formula))
+            {
+                var cuePlus = formula.IndexOf(MoleculeMassOffset.MASS_MOD_CUE_PLUS, position, StringComparison.InvariantCulture);
+                var cueMinus = formula.IndexOf(MoleculeMassOffset.MASS_MOD_CUE_MINUS, position, StringComparison.InvariantCulture);
+                if (cuePlus < 0)
+                {
+                    cuePlus = int.MaxValue;
+                }
+                if (cueMinus < 0)
+                {
+                    cueMinus = int.MaxValue;
+                }
+                // e.g. "C12H5[-1.2/1.21]", "{+1.3/1.31]",  "{-1.3]"
+                position = Math.Min(cuePlus, cueMinus);
+                var close = formula.IndexOf(']', position);
+                var parts = formula.Substring(position, close - position).Split('/');
+                double negate = Equals(cueMinus, position) ? -1 : 1;
+                if (formula.Substring(0, position).Contains(@"-"))
+                {
+                    negate *= -1;
+                }
+
+                success = double.TryParse(parts[0].Substring(2).Trim(), NumberStyles.Any, CultureInfo.InvariantCulture, out var monoMassParsed);
+                if (success)
+                {
+                    monoMassParsed *= negate;
+                    modMassMono = (modMassMono ?? 0) + monoMassParsed;
+                    var averageMassParsed = monoMassParsed;
+                    if (parts.Length > 1)
+                    {
+                        success = double.TryParse(parts[1].Trim(), NumberStyles.Any, CultureInfo.InvariantCulture, out averageMassParsed);
+                        if (success)
+                        {
+                            averageMassParsed *= negate;
+                        }
+                    }
+                    modMassAverage = (modMassAverage ?? 0) + averageMassParsed;
+                }
+                if (success)
+                {
+                    formula = formula.Substring(0, position) + formula.Substring(close + 1);
+                }
+            }
+            if (success && formula.Contains(@"/"))
+            {
+                // e.g.  "1.3/1.31"
+                var parts = formula.Split(new[] { '/' });
+                double modMassAverageParsed = 0;
+                success = double.TryParse(parts[0], NumberStyles.Any, CultureInfo.InvariantCulture, out var modMassMonoParsed) &&
```

Comment:
This appears to be the only place where we set success to false. Why not just reverse the test and throw as soon as the failure is detected? This should greatly simplify the code, because you can assume otherwise that you are parsing successfully.

Reply:
Agreed, done.